utilities/transactions/optimistic_transaction_test.cc

```diff
@@ -1687,6 +1695,108 @@ TEST_P(OptimisticTransactionTest, TimestampedSnapshotSetCommitTs) {
   ASSERT_TRUE(s.IsNotSupported());
 }
 
+TEST_P(OptimisticTransactionTest, PutEntitySuccess) {
+  constexpr char foo[] = "foo";
+  const WideColumns foo_columns{
+      {kDefaultWideColumnName, "bar"}, {"col1", "val1"}, {"col2", "val2"}};
+  const WideColumns foo_new_columns{
+      {kDefaultWideColumnName, "baz"}, {"colA", "valA"}, {"colB", "valB"}};
+
+  ASSERT_OK(txn_db->PutEntity(WriteOptions(), txn_db->DefaultColumnFamily(),
+                              foo, foo_columns));
+
+  {
+    std::unique_ptr<Transaction> txn(txn_db->BeginTransaction(WriteOptions()));
+
+    ASSERT_NE(txn, nullptr);
+    ASSERT_EQ(txn->GetNumPutEntities(), 0);
+
+    {
+      PinnableSlice value;
+      ASSERT_OK(txn->GetForUpdate(ReadOptions(), foo, &value));
+      ASSERT_EQ(value, foo_columns[0].value());
+    }
+
+    ASSERT_OK(
+        txn->PutEntity(txn_db->DefaultColumnFamily(), foo, foo_new_columns));
+
+    ASSERT_EQ(txn->GetNumPutEntities(), 1);
+
+    {
+      PinnableSlice value;
+      ASSERT_OK(txn->GetForUpdate(ReadOptions(), foo, &value));
+      ASSERT_EQ(value, foo_new_columns[0].value());
+    }
+
+    ASSERT_OK(txn->Commit());
+  }
+
+  {
+    PinnableSlice value;
+    ASSERT_OK(
+        txn_db->Get(ReadOptions(), txn_db->DefaultColumnFamily(), foo, &value));
+    ASSERT_EQ(value, foo_new_columns[0].value());
+  }
+}
+
+TEST_P(OptimisticTransactionTest, PutEntityWriteConflict) {
+  constexpr char foo[] = "foo";
+  const WideColumns foo_columns{
+      {kDefaultWideColumnName, "bar"}, {"col1", "val1"}, {"col2", "val2"}};
+  constexpr char baz[] = "baz";
+  const WideColumns baz_columns{
+      {kDefaultWideColumnName, "quux"}, {"colA", "valA"}, {"colB", "valB"}};
+
+  ASSERT_OK(txn_db->PutEntity(WriteOptions(), txn_db->DefaultColumnFamily(),
+                              foo, foo_columns));
+  ASSERT_OK(txn_db->PutEntity(WriteOptions(), txn_db->DefaultColumnFamily(),
+                              baz, baz_columns));
+
+  std::unique_ptr<Transaction> txn(txn_db->BeginTransaction(WriteOptions()));
+  ASSERT_NE(txn, nullptr);
+
+  const WideColumns foo_new_columns{{kDefaultWideColumnName, "FOO"},
+                                    {"hello", "world"}};
+  const WideColumns baz_new_columns{{kDefaultWideColumnName, "BAZ"},
+                                    {"ping", "pong"}};
+
+  ASSERT_OK(
+      txn->PutEntity(txn_db->DefaultColumnFamily(), foo, foo_new_columns));
+  ASSERT_OK(
+      txn->PutEntity(txn_db->DefaultColumnFamily(), baz, baz_new_columns));
+
+  // This PutEntity outside of a transaction will conflict with the previous
+  // write
+  const WideColumns foo_conflict_columns{{kDefaultWideColumnName, "X"},
+                                         {"conflicting", "write"}};
+  ASSERT_OK(txn_db->PutEntity(WriteOptions(), txn_db->DefaultColumnFamily(),
```

Comment:
I see the conflict being verified using one txn + one direct put without txn.

It'd be nice if we also have a test case with two txns causing the conflict